includes/embeds/class-amp-core-block-handler.php

```diff
@@ -178,4 +180,75 @@ public function ampify_cover_block( $block_content, $block ) {
 		}
 		return $block_content;
 	}
+
+	/**
+	 * Sanitize widgets that are not added via Gutenberg.
+	 *
+	 * @param Document $dom Document.
+	 */
+	public function sanitize_raw_embeds( Document $dom ) {
+		$this->process_categories_widgets( $dom );
+		$this->process_archives_widgets( $dom );
+	}
+
+	/**
+	 * Process "Categories" widgets.
+	 *
+	 * @since 1.5.2
+	 *
+	 * @param Document $dom Document.
+	 */
+	private function process_categories_widgets( Document $dom ) {
+		static $count = 0;
+
+		$widget_containers = $dom->xpath->query( '//*[ contains( @class, "widget_categories" ) ]' );
```

Comment:
This isn't matching widgets that are added via `the_widget()`. For example:

```php
$the_widget_args = array(
		'before_widget' => '<div class="widget">',
		'after_widget'  => '</div>',
		'before_title'  => '<div class="widget-title"><h3>',
		'after_title'   => '</h3></div>',
	);

the_widget( 'WP_Widget_Categories', [ 'title' => 'Categories', 'dropdown' => true ], $the_widget_args )
```

could generate:

```html
<div class="widget">
    <div class="widget-title">
        <h3>Categories</h3>
    </div>
    <form action="https://wordpressdev.lndo.site" method="get"><label class="screen-reader-text"
            for="cat">Categories</label><select name="cat" id="cat" class="postform">
            <option value="-1">Select Category</option>
            <option class="level-0" value="1">Uncategorized</option>
        </select>
    </form>
    <script>
        /* <![CDATA[ */
        (function () {
            var dropdown = document.getElementById("cat");
            function onCatChange() {
                if (dropdown.options[dropdown.selectedIndex].value > 0) {
                    dropdown.parentNode.submit();
                }
            }
            dropdown.onchange = onCatChange;
        })();
/* ]]> */
    </script>

</div>
```

Reply:
https://github.com/ampproject/amp-wp/pull/4515#discussion_r403063631

Reply:
Hardening this would be to use this query to obtain the `select`:

```
//form/select[ @name = 'cat' ]
```

I don't think it is safe to also query for `@id = "cat"` because this seems like a bug in core, where the ID needs to be auto-incremented.

And the grandparent is the widget container. Verifying it has a `script` with the expected contents should then also be done.

Reply:
See bdeb953